### PR TITLE
Issue19 theory on regression

### PR DESCRIPTION
**Pull request by CunliangGeng**
To solve #19

### CODE REVIEW

File: Theory-on-regression.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Regression Models
+
+We learned in [Introduction section](1-intro.md) that regression is to predict continuous values, and the objective of machine learning is the function or model $f(x)$, as shown in figure below.
+
+![ML-objective](image/1.1-ML-objective.png)
+
+The model $f(x)$ can also be represented as $\hat{y}(w, x)$, where $x$ is input features and $w$ is weights or parameters.
+
+Machine learning is meant to choose proper weights $w$ with [optimization algorithm](https://en.wikipedia.org/wiki/Mathematical_optimization) (learning algorithm) so that the target $\hat{y}(w, x)$ is close to real values in training examples. The "error" or "cost" between real values and predicted values is presented with [loss function](https://en.wikipedia.org/wiki/Loss_function). Then the goal of machine learning becomes minimizing the "error", i.e. loss function $J(w)$, with respect to weights $w$:
```
Comment (dafnevk) [inline Theory-on-regression.md:9]:
```suggestion
Machine learning is meant to choose proper weights $w$ with an [optimization algorithm](https://en.wikipedia.org/wiki/Mathematical_optimization) (learning algorithm) so that the target $\hat{y}(w, x)$ is close to real values in training examples. The "error" or "cost" between real values and predicted values is presented with [loss function](https://en.wikipedia.org/wiki/Loss_function). Then the goal of machine learning becomes minimizing the "error", i.e. loss function $J(w)$, with respect to weights $w$:
```

File: Theory-on-regression.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+Linear here means that the target value is expected to be a linear combination of the input features.
+
+### Model representation
+
+The model of linear regression is
+$$\hat{y}(w,x) = w_0 + w_1x_1 + w_2x_2 + \cdots + w_nx_n \$$
+where $\hat{y}$ is the target value, vector $X = (x_1, \cdots, x_n)$ are input features and $W = (w_0, \cdots, w_n)$ are weights. weight $w_0$ is also called intercept, and the other weights $(w_1, \cdots, w_n)$ are coefficients.
+
+If there is only one input feature, the model becomes
+$$\hat{y}(w,x) = w_0 + w_1x_1$$
```
Comment (dafnevk) [inline Theory-on-regression.md:27]: I would start with this, including a simple plot. This is intuitive to understand. Then make the step to multivariate linear regression.

File: Theory-on-regression.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+it's also called univariate linear regression.
+
+### Loss function
+
+Loss function is to measure the "error" between real values and predicted values. For regression, it can be presented as the following form:
```
Comment (dafnevk) [inline Theory-on-regression.md:32]: I would rephrase it a bit more entry-level
```suggestion
Suppose we have a linear model with certain values for the weights. How well does this model capture the data that we observe?

We can come up with a **loss function** that measures the "error" between real values and predicted values. For regression, we can for example calculate the squares of the differences between real and predicted points:
```

File: Theory-on-regression.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+$$ J(w) = ||y - \hat{y}||^2 $$
+
+which is the sum of squares of the differences between real values $y$ and predicted values $\hat{y}$.
+
+### Optimization algorithm
+
+Optimization algorithm is the algorithms used to minimize the loss function, i.e. to keep changing weights $w$ to reduce loss $J(w)$ until it hopefully ends up at a minimum.
```
Comment (dafnevk) [inline Theory-on-regression.md:40]: Some rephrasing in simpler language:
```suggestion
Now that we have defined a loss function, we want to choose the weights so that the loss is as small as possible. We do this with an **optimization algorithm**: an algorithm used to minimize the loss function, i.e. to keep changing weights $w$ to reduce loss $J(w)$ until it hopefully ends up at a minimum.
```

File: Theory-on-regression.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+The most basic and popular optimization algorithm is [gradient descent](https://en.wikipedia.org/wiki/Gradient_descent), also known as steepest descent. It can be described as below:
+
+$$ \color{gray} \text{repeat until convergence: } \\
```
Comment (dafnevk) [inline Theory-on-regression.md:44]: I think the mathematics can be skipped here. Just showing the illustrations is much more powerful, if learners are interested we can refer them to reading material on this.

File: Theory-on-regression.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+The process of gradient descent for two features would look like the dynamic diagram below:
+![Gradient descent 2D dynamic](image/Gradient_descent_dynamic.gif)
+<!-- Figure Gradient_descent_dynamic.gif from https://blog.paperspace.com/intro-to-optimization-in-deep-learning-gradient-descent/ -->
+
+
+## Neural network (non-linear regression)
+
+### Model representation
+
```
Comment (dafnevk) [inline Theory-on-regression.md:64]: Here we can borrow from the Deep Learning lesson, see [this part](https://github.com/carpentries-incubator/deep-learning-intro/blob/gh-pages/_episodes/01-introduction.md#neural-networks). I think we can copy most of it, I find the illustration of a single neuron also very useful.

File: Theory-on-regression.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+<!-- Text from https://scikit-learn.org/stable/modules/neural_networks_supervised.html#multi-layer-perceptron -->
+The input layer consists of a set of neurons representing the input features.
+
+ Each neuron in the hidden layer transforms the values from the previous layer with a weighted linear summation $w_1x_1 + w_2x_2 + \cdots + w_nx_n + bias$, followed by a non-linear activation function, e.g. the [logistic or sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function).
+
+ The output layer receives the values from the last hidden layer and transforms them into output values with linear function.
+
+The neural network available in `scikit-learn` is just look like the architecture in the diagram, i.e. multi-layer perceptron. To use more complex neural network models, deep learning framework should be used, e.g. [PyTorch, Keras, TensorFlow, etc](https://en.wikipedia.org/wiki/Comparison_of_deep_learning_software). Want to have a look at various neural network models? Try https://alexlenail.me/NN-SVG/index.html.
+
+#### Activation function
```
Comment (dafnevk) [inline Theory-on-regression.md:78]: I don't think it's very relevant to show the different activation functions, we can simply note that we need a non-linear function there and what is the most common choice.

File: Theory-on-regression.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+[Activation function](https://en.wikipedia.org/wiki/Activation_function) is also called transfer function. Various activation functions are available in `scikit-learn`, e.g. identity, logistic, tanh and relu. Take a look at [this table](https://en.wikipedia.org/wiki/Activation_function#Table_of_activation_functions) to know more about these functions.
+
+![Activation functions](image/Activation_functions.png)
+<!-- Table is from https://en.wikipedia.org/wiki/Activation_function#Table_of_activation_functions -->
+
+
+### Loss function
+Loss function can be shared by all regression models, so loss function $J(w) = ||y - \hat{y}||^2$ is also used for neural network model.
+
+### Optimization algorithms
```
Comment (dafnevk) [inline Theory-on-regression.md:88]: Here I also don't think you have to go into too much detail about different optimizers. It could be useful to show an image of backpropegation.

File: Theory-on-regression.md (hunk 7 above)
Comment (dafnevk) [reply Theory-on-regression.md:78]: It's interesting to add here what the benefits and drawbacks are of neural networks over linear regression. E.g.:

`The design of the neural network, with it's non-linearity and layers stacked on top of each other, allows for much more complex patterns to be detected than with linear regression. We can view the neurons as 'feature detectors', that retrieve relevant pieces of information from the input data. The flipside of the coin, is that often we need large amounts of training data to be able to learn these features.`

File: Theory-on-regression.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (CunliangGeng) [reply Theory-on-regression.md:9]: Fully agree. Created an issue for this https://github.com/esciencecenter-digital-skills/SICSS-odissei-machine-learning/issues/41

File: Theory-on-regression.md (hunk 2 above)
Comment (CunliangGeng) [reply Theory-on-regression.md]: Good idea, updated as what you suggested.

File: Theory-on-regression.md (hunk 5 above)
Comment (CunliangGeng) [reply Theory-on-regression.md:54]: Thanks, I moved the math to the end of this section as a reference. We don't have to use it in the slides, but it's nice to keep in the full material.

File: Theory-on-regression.md (hunk 7 above)
Comment (CunliangGeng) [reply Theory-on-regression.md:90]: > I don't think it's very relevant to show the different activation functions, we can simply note that we need a non-linear function there and what is the most common choice.

Not fully agree. `scikit-learn` has the parameter to set activation function, it's nice to briefly  mention these functions here, of course we don't want to go details. And you're right, the most important is let them know the most common choice.

Rephrased text a bit.

File: Theory-on-regression.md (hunk 8 above)
Comment (CunliangGeng) [reply Theory-on-regression.md:100]: Yes, same as the activation function, we won't mention too much detail about optimisers, but the commonly used choice.
As for the backprop, I even don't want to mention it actually. No more image, otherwise we have to explain more...

**Comment by CunliangGeng** [comment]
Thanks for the comments @dafnevk. Most of them are accepted or text updated.

**Comment by CunliangGeng** [comment]
@dsmits The PR has been updated following Dafne's comments, could you also have a look?

**Comment by dafnevk** [comment]
Thanks Cunliang, looks good to me now! (But indeed would be good if Djura also has a look)

**Review by dsmits — APPROVED** [review]
Great work!
